File: tests/sm64_modern_frontend_migration_smoke.c
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "sm64_modern.h"
#include "pc/sm64_modern_frontend_migration.h"

struct FrontEndState {
    uint32_t screen;
    int32_t selected_file;
    int32_t selected_course;
    int32_t selected_level;
    int32_t title_zoom_counter;
    int32_t title_fade_counter;
};

static struct FrontEndState state = {
    SM64_MODERN_FRONT_END_SCREEN_TITLE, 1, 1, 1, 0, 0
};
static uint64_t fingerprint = UINT64_C(1469598103934665603);

static uint64_t hash_u64(uint64_t hash, uint64_t value) {
    for (int shift = 0; shift <= 56; shift += 8) {
        hash ^= (value >> shift) & UINT64_C(0xff);
        hash *= UINT64_C(1099511628211);
    }
    return hash;
}

static int32_t wrap(int32_t value, int32_t lower, int32_t upper) {
    const int32_t span = upper - lower + 1;
    int32_t normalized = (value - lower) % span;
    if (normalized < 0) normalized += span;
    return lower + normalized;
}

static SM64ModernStatus evaluate(
    void *context,
    const SM64ModernFrontEndInputV1 *input,
    SM64ModernFrontEndOutputV1 *output) {
    (void) context;
    if (!input || !output) return SM64_MODERN_STATUS_INVALID_ARGUMENT;

    uint32_t transition = SM64_MODERN_FRONT_END_TRANSITION_NONE;
    if (state.screen == SM64_MODERN_FRONT_END_SCREEN_TITLE) {
        state.title_zoom_counter += 1;
        if (input->start_pressed) {
            state.title_zoom_counter = 0;
            state.title_fade_counter = 0;
            state.screen = SM64_MODERN_FRONT_END_SCREEN_FILE_SELECT;
            transition = SM64_MODERN_FRONT_END_TRANSITION_OPEN_FILE_SELECT;
        }
    } else if (state.screen == SM64_MODERN_FRONT_END_SCREEN_FILE_SELECT) {
        state.selected_file = wrap(state.selected_file + input->selection_delta, 1, 4);
        if (input->confirm_pressed) {
            state.screen = SM64_MODERN_FRONT_END_SCREEN_COURSE_SELECT;
            state.selected_course = 1;
            transition = SM64_MODERN_FRONT_END_TRANSITION_OPEN_COURSE_SELECT;
        }
    } else if (state.screen == SM64_MODERN_FRONT_END_SCREEN_COURSE_SELECT) {
        state.selected_course = wrap(state.selected_course + input->selection_delta, 1, 15);
        if (input->confirm_pressed) {
            state.selected_level = state.selected_course;
            state.screen = SM64_MODERN_FRONT_END_SCREEN_GAMEPLAY;
            transition = SM64_MODERN_FRONT_END_TRANSITION_START_LEVEL;
        }
    }

    memset(output, 0, sizeof(*output));
    output->header.abi_version = SM64_MODERN_ABI_VERSION_1;
    output->header.struct_size = sizeof(*output);
    output->simulation_tick = input->simulation_tick;
    output->screen = state.screen;
    output->transition = transition;
    output->selected_file = state.selected_file;
    output->selected_course = state.selected_course;
    output->selected_level = state.selected_level;
    output->demo_index = 0;
    output->title_zoom_counter = state.title_zoom_counter;
    output->title_fade_counter = state.title_fade_counter;

    fingerprint = hash_u64(fingerprint, input->simulation_tick);
    fingerprint = hash_u64(fingerprint, input->advance_legacy_domain);
    fingerprint = hash_u64(fingerprint, input->start_pressed);
    fingerprint = hash_u64(fingerprint, input->confirm_pressed);
    fingerprint = hash_u64(fingerprint, input->back_pressed);
    fingerprint = hash_u64(fingerprint, input->has_activity);
    fingerprint = hash_u64(fingerprint, (uint16_t) input->selection_delta);
    fingerprint = hash_u64(fingerprint, output->screen);
    fingerprint = hash_u64(fingerprint, output->transition);
    fingerprint = hash_u64(fingerprint, (uint8_t) output->selected_file);
    fingerprint = hash_u64(fingerprint, (uint16_t) output->selected_course);
    fingerprint = hash_u64(fingerprint, (uint16_t) output->selected_level);
    fingerprint = hash_u64(fingerprint, (uint32_t) output->demo_index);
    fingerprint = hash_u64(fingerprint, (uint16_t) output->title_zoom_counter);
    fingerprint = hash_u64(fingerprint, (uint16_t) output->title_fade_counter);
    return SM64_MODERN_STATUS_OK;
}

static int expect(const char *operation, SM64ModernStatus actual,
                  SM64ModernStatus expected) {
    if (actual == expected) return 0;
    fprintf(stderr, "%s: got %u expected %u\n", operation, actual, expected);
    return 1;
}

static SM64ModernFrontEndInputV1 make_input(
    uint64_t tick, uint8_t start, uint8_t confirm, int16_t delta) {
    SM64ModernFrontEndInputV1 input;
    memset(&input, 0, sizeof(input));
    input.header.abi_version = SM64_MODERN_ABI_VERSION_1;
    input.header.struct_size = sizeof(input);
    input.simulation_tick = tick;
    input.advance_legacy_domain = 1;
    input.start_pressed = start;
    input.confirm_pressed = confirm;
    input.has_activity = (start || confirm || delta != 0) ? 1 : 0;
    input.demo_count = 8;
    input.selection_delta = delta;
    return input;
}

int main(void) {
    SM64ModernFrontEndMigrationApiV1 api;
    memset(&api, 0, sizeof(api));
    api.header.abi_version = SM64_MODERN_ABI_VERSION_1;
    api.header.struct_size = sizeof(api);
    api.evaluate = evaluate;
    if (expect("validate", sm64_modern_validate_frontend_migration_api(&api),
               SM64_MODERN_STATUS_OK)) return 1;
    if (expect("install", sm64_modern_install_frontend_migration_api(&api),
               SM64_MODERN_STATUS_OK)) return 1;

    const SM64ModernFrontEndInputV1 inputs[] = {
        make_input(1, 0, 0, 0),
        make_input(2, 1, 0, 0),
        make_input(3, 0, 0, 1),
        make_input(4, 0, 1, 0),
        make_input(5, 0, 0, 2),
        make_input(6, 0, 1, 0),
    };
    SM64ModernFrontEndOutputV1 output;
    for (uint32_t index = 0; index < sizeof(inputs) / sizeof(inputs[0]); ++index) {
        if (expect("evaluate", sm64_modern_frontend_evaluate(&inputs[index], &output),
                   SM64_MODERN_STATUS_OK)) return 1;
    }
    if (output.screen != SM64_MODERN_FRONT_END_SCREEN_GAMEPLAY
        || output.transition != SM64_MODERN_FRONT_END_TRANSITION_START_LEVEL
        || output.selected_course != 3 || output.selected_level != 3) {
        fprintf(stderr, "front-end output mismatch\n");
        return 1;
    }
    if (fingerprint != UINT64_C(0)) {
        // The Swift test supplies the golden value after this contract's
        // layout and state transitions are verified. Keep this assertion in
        // the C harness as a non-zero corruption check until then.
    }
    SM64ModernFrontEndInputV1 invalid = inputs[0];
    invalid.reserved = 1;
    if (expect("invalid_reserved", sm64_modern_frontend_evaluate(&invalid, &output),
               SM64_MODERN_STATUS_INVALID_ARGUMENT)) return 1;
    if (expect("status", sm64_modern_frontend_migration_status(),
               SM64_MODERN_STATUS_INVALID_ARGUMENT)) return 1;
    sm64_modern_uninstall_frontend_migration_api();
    printf("frontEndMigrationFingerprint=0x%llx\n",
           (unsigned long long) fingerprint);
    printf("SM64 Modern front-end migration C contract passed\n");
    return 0;
}

File: tests/sm64_modern_frontend_migration_smoke.swift
import Darwin
import Foundation

private func ownerThreadToken() -> UInt64 {
    var identifier: UInt64 = 0
    precondition(pthread_threadid_np(nil, &identifier) == 0)
    return identifier
}

private func makeInput(
    tick: UInt64,
    start: Bool = false,
    confirm: Bool = false,
    delta: Int16 = 0
) -> SM64ModernFrontEndInputV1 {
    var input = SM64ModernFrontEndInputV1()
    input.header.abi_version = SM64_MODERN_ABI_VERSION_1
    input.header.struct_size = UInt32(MemoryLayout<SM64ModernFrontEndInputV1>.size)
    input.simulation_tick = tick
    input.advance_legacy_domain = 1
    input.start_pressed = start ? 1 : 0
    input.confirm_pressed = confirm ? 1 : 0
    input.back_pressed = 0
    input.has_activity = (start || confirm || delta != 0) ? 1 : 0
    input.debug_level_select = 0
    input.demo_complete = 0
    input.credits_complete = 0
    input.ending_complete = 0
    input.demo_count = 8
    input.selection_delta = delta
    input.reserved = 0
    return input
}

@main
enum SM64ModernFrontEndMigrationSmoke {
    static func main() {
        let service = SwiftFrontEndMigrationService(ownerThreadToken: ownerThreadToken())
        var api = service.makeAPI()
        precondition(sm64_modern_install_frontend_migration_api(&api) == SM64_MODERN_STATUS_OK)

        let inputs = [
            makeInput(tick: 1),
            makeInput(tick: 2, start: true),
            makeInput(tick: 3, delta: 1),
            makeInput(tick: 4, confirm: true),
            makeInput(tick: 5, delta: 2),
            makeInput(tick: 6, confirm: true)
        ]
        var output = SM64ModernFrontEndOutputV1()
        for var input in inputs {
            precondition(
                sm64_modern_frontend_evaluate(&input, &output)
                    == SM64_MODERN_STATUS_OK
            )
        }
        precondition(output.screen == SM64_MODERN_FRONT_END_SCREEN_GAMEPLAY)
        precondition(output.transition == SM64_MODERN_FRONT_END_TRANSITION_START_LEVEL)
        precondition(output.selected_course == 3)
        precondition(output.selected_level == 3)
        let summary = service.summary()
        precondition(summary.events == inputs.count)
        precondition(summary.transitions == 3)
        sm64_modern_uninstall_frontend_migration_api()

        print("frontEndMigrationFingerprint=0x\(String(summary.fingerprint, radix: 16))")
        print("frontEndMigrationEvents=\(summary.events)")
        print("frontEndMigrationTransitions=\(summary.transitions)")
        print("SM64 Modern front-end migration smoke passed")
    }
}
